File: go/canonical.go
package sesame

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

// BodyHashHex returns the lowercase-hex SHA-256 of the (possibly-encrypted) body.
func BodyHashHex(body []byte) string {
	sum := sha256.Sum256(body)
	return hex.EncodeToString(sum[:])
}

// RequestCanonical builds the Tier 1 request canonical string:
// method LF path LF timestamp LF nonce LF body-hash [LF scope].
// scope is nil unless Tier 2 is active.
func RequestCanonical(method, path, timestamp, nonce, bodyHashHex string, scope *string) string {
	fields := []string{method, path, timestamp, nonce, bodyHashHex}
	if scope != nil {
		fields = append(fields, *scope)
	}
	return strings.Join(fields, "\n")
}

// ResponseCanonical builds the response canonical string:
// "RESPONSE" LF correlation LF timestamp LF nonce LF body-hash [LF scope].
func ResponseCanonical(correlation, timestamp, nonce, bodyHashHex string, scope *string) string {
	fields := []string{"RESPONSE", correlation, timestamp, nonce, bodyHashHex}
	if scope != nil {
		fields = append(fields, *scope)
	}
	return strings.Join(fields, "\n")
}

File: python/src/sesame/canonical.py
"""SESAME canonical strings (draft v0.5).

Byte-identical to the Rust crate and C++ SDK: the same fields, the same LF
separators, no trailing newline.
"""

from __future__ import annotations

import hashlib
from typing import Optional


def body_hash_hex(body: bytes) -> str:
    """Lowercase-hex SHA-256 of the (possibly-encrypted) body."""
    return hashlib.sha256(body).hexdigest()


def request_canonical(
    method: str,
    path: str,
    timestamp: str,
    nonce: str,
    body_hash_hex: str,
    scope: Optional[str] = None,
) -> str:
    """method LF path LF timestamp LF nonce LF body-hash [LF scope]."""
    fields = [method, path, timestamp, nonce, body_hash_hex]
    if scope is not None:
        fields.append(scope)
    return "\n".join(fields)


def response_canonical(
    correlation: str,
    timestamp: str,
    nonce: str,
    body_hash_hex: str,
    scope: Optional[str] = None,
) -> str:
    """"RESPONSE" LF correlation LF timestamp LF nonce LF body-hash [LF scope]."""
    fields = ["RESPONSE", correlation, timestamp, nonce, body_hash_hex]
    if scope is not None:
        fields.append(scope)
    return "\n".join(fields)
